### PR TITLE
provider/aws: Add FlowLog resource

### PR DESCRIPTION
**Pull request by catsby**
This is a proposal config/work for adding a [FlowLog][1] resource.

**Syntax:**

```javascript
resource "aws_flow_log" "test_flow_log" {
  log_group_name = "tf-test-log-group"
  iam_role_arn = "arn:aws:iam::470663696735:role/tf-test-cloud"
  vpc_id = "${aws_vpc.default.id}"
  traffic_type = "ALL"
}
```

**API Docs:**

- [CreateFlowLogs](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateFlowLogs.html)
- [Describe FlowLogs](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeFlowLogs.html)
- [Delete FlowLogs](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DeleteFlowLogs.html)
- There is no `Update` or `Modify`, things will be `ForceNew`. 

A `FlowLog` can be attached to either a VPC, Subnet, or ENI. The API call is the same, you just supply the resource `id` and `type`. Here I propose separate `vpc_id`, `subnet_id`, and `eni_id` syntax, to avoid a user having to specify both a resource `id` and the type. 

Additionally, the API supports creating multiple flows in a single call, but this creates multiple `FlowLog`s, , so our implementation will be 1:1. 

**Remaining:**

- [x] Docs
- [x] Implementation
- [x] Acceptance tests

[1]: https://aws.amazon.com/blogs/aws/vpc-flow-logs-log-and-view-network-traffic-flows

### CODE REVIEW

File: builtin/providers/aws/provider.go [hunk 1]
```diff
@@ -85,12 +85,13 @@ func Provider() terraform.ResourceProvider {
 		ResourcesMap: map[string]*schema.Resource{
 			"aws_app_cookie_stickiness_policy": resourceAwsAppCookieStickinessPolicy(),
 			"aws_autoscaling_group":            resourceAwsAutoscalingGroup(),
+			"aws_flow_log":                     resourceAwsFlowLog(),
```
Comment (phinze) [inline builtin/providers/aws/provider.go]: Alphabetize por favor! :capital_abcd: :grin:

**Comment by catsby** [comment]
cc @radeksimko since I know he has at least looked into Flow Logs

File: builtin/providers/aws/resource_aws_flow_log.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	})
+
+	if err != nil {
+		return fmt.Errorf("[WARN] Error deleting Flow Log with ID (%s), error: %s", d.Id(), err)
+	}
+
+	return nil
+}
+
+func flowLogStateRefreshFunc(conn *ec2.EC2, sn string) resource.StateRefreshFunc {
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_flow_log.go]: Looks like this guy is unused.

File: builtin/providers/aws/resource_aws_flow_log.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	opts := &ec2.CreateFlowLogsInput{
+		DeliverLogsPermissionARN: aws.String(d.Get("iam_role_arn").(string)),
+		LogGroupName:             aws.String(d.Get("log_group_name").(string)),
+		ResourceIDs:              []*string{aws.String(resourceId)},
+		ResourceType:             aws.String(resourceType),
+		TrafficType:              aws.String(d.Get("traffic_type").(string)),
+	}
+	resp, err := conn.CreateFlowLogs(opts)
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_flow_log.go]: Log the create with the `awsutil.StringValue` of `opts` here?

File: builtin/providers/aws/resource_aws_flow_log.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	d.Set("iam_role_arn", fl.DeliverLogsPermissionARN)
+	d.Set("flow_log_status", fl.FlowLogStatus)
+	d.Set("deliver_log_status", fl.DeliverLogsStatus)
+
+	return nil
+}
+
+func resourceAwsLogFlowDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	_, err := conn.DeleteFlowLogs(&ec2.DeleteFlowLogsInput{
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_flow_log.go]: Log the delete here?

File: builtin/providers/aws/resource_aws_flow_log_test.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_flow_log" "test_flow_log_subnet" {
+        # log_group_name needs to exist before hand
+        # until we have a CloudWatch Log Group Resource
+        log_group_name = "tf-test-log-group"
```
Comment (phinze) [inline builtin/providers/aws/resource_aws_flow_log_test.go]: Can you interpolate an env variable for this and fail if it's not set? That's how we've handled required prerequisites elsewhere. For example resource_dme_record_test.

**Comment by phinze** [comment]
Review complete - Looking great overall!

File: builtin/providers/aws/resource_aws_flow_log_test.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/aws-sdk-go/aws"
```
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_flow_log_test.go]: why this import?

Comment (catsby) [inline builtin/providers/aws/resource_aws_flow_log_test.go]: Wow, great catch! My editor must have auto populated that :scream: 

Fixed in 0bf127a

File: builtin/providers/aws/resource_aws_flow_log_test.go (hunk 5 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_flow_log_test.go]: Done in 0bf127a

File: builtin/providers/aws/resource_aws_flow_log.go (hunk 4 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_flow_log.go]: Done in 285a88b

File: builtin/providers/aws/resource_aws_flow_log.go (hunk 3 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_flow_log.go]: Done in 285a88b

File: builtin/providers/aws/resource_aws_flow_log.go (hunk 2 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_flow_log.go]: :scream:

**Comment by catsby** [comment]
@phinze cleaned up, sorry for the sloppy 😐

File: builtin/providers/aws/resource_aws_flow_log.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	opts := &ec2.CreateFlowLogsInput{
+		DeliverLogsPermissionARN: aws.String(d.Get("iam_role_arn").(string)),
+		LogGroupName:             aws.String(d.Get("log_group_name").(string)),
+		ResourceIDs:              []*string{aws.String(resourceId)},
+		ResourceType:             aws.String(resourceType),
+		TrafficType:              aws.String(d.Get("traffic_type").(string)),
+	}
+
+	log.Printf(
+		"[DEBUG] Flow Log Create configuration: %s", awsutil.StringValue(opts))
```
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_flow_log.go]: Nice! :+1:  Isn't this basically addressing https://github.com/hashicorp/terraform/issues/2179 ?

Comment (catsby) [inline builtin/providers/aws/resource_aws_flow_log.go]: `awsutil.StringValue` does, yes, but to close #2179 we need to use it in All The Places™

File: builtin/providers/aws/resource_aws_flow_log_test.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+      ]
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_flow_log" "test_flow_log" {
+        # log_group_name needs to exist before hand
+        # until we have a CloudWatch Log Group Resource
```
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_flow_log_test.go]: If we plan to release this without the CloudWatch Log Group resource, there should IMO be a warning/note in docs mentioning it.

File: website/source/docs/providers/aws/r/flow_log.html.markdown [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  traffic_type = "ALL"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+*	`log_group_name` - (Required) The name of the CloudWatch log group
+*	`iam_role_arn` - (Required) The ARN for the IAM role that's used to post flow
```
Comment (radeksimko) [inline website/source/docs/providers/aws/r/flow_log.html.markdown]: Strange indentions here.

**Comment by radeksimko** [comment]
@catsby It looks ok, just one question here:
What is the benefit behind saving `flow_log_status` and `deliver_log_status`?

I thought that change in the _state_ should be changed because user/human has done something in order to change the state. My understanding is that these two things will change without user's interaction over time and so far I don't see how would any other Terraform resource react on that.

**Comment by radeksimko** [comment]
I reckon this can be fixed later on after merging, but just FYI regarding example in docs:

http://docs.aws.amazon.com/cli/latest/reference/ec2/create-flow-logs.html

> In your request, you must also specify an IAM role that has permission to publish logs to CloudWatch Logs.

We're not defining that permission anywhere.

File: website/source/docs/providers/aws/r/flow_log.html.markdown (hunk 9 above)
Comment (catsby) [inline website/source/docs/providers/aws/r/flow_log.html.markdown]: cleaned up in d5f962b

**Comment by catsby** [comment]
@radeksimko Thanks for reviewing

> What is the benefit behind saving flow_log_status and deliver_log_status?

I thought it would be useful, but in hindsight agree, so I removed them in 87c7f63

> We're not defining that permission anywhere.

I updated the documentation to include the needed IAM things to get Flow Logs working, and verified they work. The Logs must Flow

File: builtin/providers/aws/resource_aws_flow_log_test.go (hunk 8 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_flow_log_test.go]: I updated the docs to mention that it must exist already. I did see your Log Group PR too (thanks!), which will help when that gets completed

**Comment by phinze** [comment]
LGTM

**Pull request merged**